### ISSUE TITLE
DB: Uniqueness constraint in "waveforms_continuoustrace" needs to be extended

### ISSUE THREAD

**Issue by megies** [krischer/jane#84]
In real life, it unfortunately sometimes happens that data is duplicated in local files, e.g. mostly identical miniseed packets (maybe with just a different sequence number) duplicated in the same file. I know this can happen e.g. when accidentally running two `slarchive` processes (to save data from seedlink buffer to disk) where only a single one should run.

Example:
```
In [1]: st = read('Z3.A104A..HHE.D.2018.042')

In [2]: st
Out[2]: 
2 Trace(s) in Stream:
Z3.A104A..HHE | 2018-02-11T00:00:00.400000Z - 2018-02-12T00:00:00.340000Z | 100.0 Hz, 8639995 samples
Z3.A104A..HHE | 2018-02-11T00:00:00.400000Z - 2018-02-12T00:00:00.340000Z | 100.0 Hz, 8639995 samples
```

Right now, jane can not index such files (if at least one contiguous data chunk in the file is fully duplicated), because during indexing a unique constraint on `waveforms_contiguoustrace` gets violated:

```
2018-07-26 08:56:54 UTC [10285-70] jane@jane ERROR:  duplicate key value violates unique constraint "waveforms_continuoustrace_file_id_1a3787d9_uniq"
2018-07-26 08:56:54 UTC [10285-71] jane@jane DETAIL:  Key (file_id, network, station, location, channel, timerange)=(15868162, Z3, A104A, , HHE, ["2018-04-10 00:00:00.45+00","2018-04-11 00:00:03.4+00")) already exists.
```

To add to initial confusion, the conflicting and supposedly "existing" database entry mentioned in the postgres error (here some file with id 15868162 in `waveforms_file`) can not even be found in the database, because the conflict happens during one transaction that is then never committed but instantly rolled back.

Two problematic outcomes of this are:
 - postgres error log blows up at enormous speed (if indexer is looping), we had it fill up the whole partition (which can lead to even nastier problems in consequence)
 - id field in `waveforms_contiguoustrace` gets blown up (incremented by entries that are immediately rolled back and that never make it into the table actually) which might lead to further problems when hitting the maximum possible number of the id field (if this problem is not detected soon enough)

---

Two implications:
 - [x] **must:** unique constraint on `waveforms_contiguoustrace` *needs to be changed*, it has to include `pos` field (which I assume gives index of Trace when reading the file into a Stream in ObsPy). then jane should be able to index such problematic files
 - [ ] **"would be nice"**: ideally, the indexer *should* remember what files could not be indexed and when starting another loop should skip those files if they did not change in the meantime (checking modification time)

**Comment by barsch** [comment]
https://docs.djangoproject.com/en/1.11/topics/migrations/

basically change your model - after changes are done run ```python manage.py makemigrations [module]``` (module is optional) - after that apply your migrations via ```python manage.py migrate```

**Comment by barsch** [comment]
in order to set the unique constraint you also need to use ```unique_together``` in the Meta class, see

https://stackoverflow.com/questions/2201598/how-to-define-two-fields-unique-as-couple

**Comment by megies** [comment]
And django knows which migrations were already acted on because it stores that info in the DB in "django_migrations" table, I guess?

> in order to set the unique constraint you also need to use unique_together in the Meta class, see

Mhm.. that's the spot in source code I saw will need changing

So..  django creates the migration automatically?

 - in jane source code in a PR, change model in `jane/waveforms/models.py` to include `pos` in `unique_together`
 - on production server..
   - stop jane
   - run `python manage.py makemigrations`
   - run `python manage.py migrate`
   - start jane

?

**Comment by megies** [comment]
> -> so essentiall generating migrations should be done on development server, on productive you only apply the migrations ....

alright.. fingers crossed :see_no_evil:

**Comment by megies** [comment]
Closed by 956cf03a90a993a6b872e8c3e412acdd268df482 and fccfcaf8f92bb86f5baa630153a5b0824cdd1e6f
